### PR TITLE
nsexec

### PR DESCRIPTION
**Pull request by nayuta723**
## Implementation Summary
I implemented [a clone_parent function](https://github.com/nayuta-ai/docker_internal/blob/nsexec/clone_parent/src/nsexec.c) and [a test file corresponding to the clone_parent](https://github.com/nayuta-ai/docker_internal/blob/nsexec/clone_parent/tests/C/main.cpp).

## Scope of Impact
The clone_parent creates a child process, and both processes continuously run.

## Particular points to check
I would appreciate your comments on whether there is any information missing about README or the commits.

## Test
We can test each function by using the following command:
```
$ git clone https://github.com/nayuta-ai/docker_internal.git
$ cd docker_internal
$ git checkout nsexec/clone_parent
$ make build
$ make run
$ make exec
$ cd tests/C
$ cmake .
$ make
$ ./example_test
```

### CODE REVIEW

File: doc/nsenter.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+In the STAGE_CHILD case, the child process opens the sync_pipe for reading and closes the end for writing. The child process creates a new user namespace using the unshare system call with the CLONE_NEWUSER flag, which allows the child to have a separate user and group ID namespace from its parent. The child process then writes its process ID to the sync_pipe and terminates.
+
+The EXPECT_GE and EXPECT_EQ macros are used to verify the results of the test and report failure if any of the expectations are not met. The exit function is used to exit the child process with a status of 0.
+
+## How to test
+```
+$ git clone https://github.com/nayuta-ai/docker_internal.git
+$ cd docker_internal
+$ git checkout nsexec/clone_parent
```
Comment (ryuji0123) [inline doc/nsenter.md:37]: git checkout nsexec

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,3 @@
+module docker_internal
```
Comment (ryuji0123) [inline go.mod]: Do we need this file in this PR?

File: src/nsexec.c [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+/* Assume the stack grows down, so arguments should be above it. */
+struct clone_t {
+	/*
+	 * Reserve some space for clone() to locate arguments
+	 * and retcode in this place
+	 */
+	char stack[4096] __attribute__((aligned(16)));
+	char stack_ptr[0];
+
+	/* There's two children. This is used to execute the different code. */
```
Comment (ryuji0123) [inline src/nsexec.c]: Maybe
This is -> They are

**Review by ryuji0123 — CHANGES_REQUESTED** [review]
Basically, your code seems fine. Please check my comments.

File: doc/nsenter.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+$ make build
+$ make run
+$ make exec
+$ cd tests/C
+$ cmake .
+$ make
+$ ./example_test
```
Comment (ryuji0123) [inline doc/nsenter.md]: example_test -> nsexec_test

**Review by ryuji0123 — CHANGES_REQUESTED** [review]

File: tests/C/main.cpp [hunk 5]
```diff
@@ -1,5 +1,60 @@
 #include <gtest/gtest.h>
```
Comment (ryuji0123) [inline tests/C/main.cpp]: Let's follow [this style](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes).

**Review by ryuji0123 — CHANGES_REQUESTED** [review]

File: doc/nsenter.md (hunk 1 above)
Comment (nayuta723) [reply doc/nsenter.md:37]: I used nsexec/clone_parent tag for switching source code. Is it preferable to change the source code on a per-branch basis?

**Comment by nayuta723** [comment]
@ryuji0123 
I modified the previous one to the directory structure, which summarizes the "src" and "test" directories in one director. But the different points are no longer visible. I know this is hard to review, but please comment on that.

**Comment by ryuji0123** [comment]
@nayuta-ai 
As I told you before, you can resolve conversations first, to show what you've modified.

File: .gitignore [hunk 6]
```diff
@@ -65,21 +65,16 @@ dkms.conf
...
 # Executables
 
-# Google Test output files
-CMakeFiles
-cmake_install.cmake
-CMakeCache.txt
-example_test
-tests/*/Makefile
+# Directories
+build
```
Comment (ryuji0123) [inline .gitignore]: "build/" and "build" is different. This line should be "build/". You should also check the difference.

File: doc/nsenter.md (hunk 1 above)
Comment (ryuji0123) [reply doc/nsenter.md:37]: I see. In that case, we should fetch the tag before checking out.

File: nsexec/src/nsexec.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#ifndef NSEXEC_H
+#define NSEXEC_H
+
+#include <setjmp.h>
+
+/* Assume the stack grows down, so arguments should be above it. */
+struct clone_t {
+	/*
```
Comment (ryuji0123) [inline nsexec/src/nsexec.h]: About indent spaces, you can see [this](https://google.github.io/styleguide/cppguide.html#Spaces_vs._Tabs). It is also the same case with other files.

**Review by ryuji0123 — CHANGES_REQUESTED** [review]

File: nsexec/src/nsexec.h [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+   */
+  char stack[4096] __attribute__((aligned(16)));
+  char stack_ptr[0];
+
+  /* There's two children. They are used to execute the different code. */
+  jmp_buf *env;
+  int jmpval;
+};
+
+static int child_func(void *arg)
```
Comment (ryuji0123) [inline nsexec/src/nsexec.h]: https://github.com/okuda-seminar/Engineering/issues/661#issuecomment-1434276807

**Review by ryuji0123 — CHANGES_REQUESTED** [review]

**Review by ryuji0123 — APPROVED** [review]
LGTM

**Pull request merged**